File: backend/src/modules/equipment/equipment.service.js
const prisma = require('../../lib/prisma');
const { HttpError } = require('../../lib/httpError');

async function getFloorWithContext(floorId) {
  const floor = await prisma.floor.findUnique({
    where: { id: floorId },
    include: { building: true },
  });
  if (!floor) {
    throw new HttpError(404, 'Floor not found');
  }
  return floor;
}

async function listEquipmentByFloor(floorId) {
  await getFloorWithContext(floorId);
  return prisma.equipment.findMany({
    where: { floorId },
    orderBy: { name: 'asc' },
    include: { _count: { select: { points: true } } },
  });
}

async function listEquipmentBySite(siteId) {
  const site = await prisma.site.findUnique({ where: { id: siteId } });
  if (!site) {
    throw new HttpError(404, 'Site not found');
  }
  return prisma.equipment.findMany({
    where: { siteId },
    include: {
      building: { select: { id: true, name: true } },
      floor: { select: { id: true, name: true } },
      _count: { select: { points: true } },
    },
    orderBy: [{ buildingId: 'asc' }, { floorId: 'asc' }, { name: 'asc' }],
  });
}

async function createEquipment(floorId, data) {
  const floor = await getFloorWithContext(floorId);
  const siteId = floor.building.siteId;
  const buildingId = floor.buildingId;

  const { name, code, equipmentType, status, templateName, address, instanceNumber } = data;
  if (!name || !code || !equipmentType) {
    throw new HttpError(400, 'name, code, and equipmentType are required');
  }

  const template =
    templateName != null && String(templateName).trim() !== ''
      ? String(templateName).trim()
      : null;

  const inst =
    instanceNumber != null && String(instanceNumber).trim() !== ''
      ? String(instanceNumber).trim()
      : null;

  return prisma.equipment.create({
    data: {
      siteId,
      buildingId,
      floorId,
      name: String(name).trim(),
      code: String(code).trim(),
      equipmentType: String(equipmentType).trim(),
      ...(template ? { templateName: template } : {}),
      ...(status ? { status } : {}),
      ...(address != null && String(address).trim() !== ''
        ? { address: String(address).trim() }
        : {}),
      ...(inst ? { instanceNumber: inst } : {}),
    },
  });
}

async function getEquipmentById(id) {
  const equipment = await prisma.equipment.findUnique({
    where: { id },
    include: {
      site: true,
      building: true,
      floor: true,
    },
  });
  if (!equipment) {
    throw new HttpError(404, 'Equipment not found');
  }
  return equipment;
}

async function updateEquipment(id, data) {
  await getEquipmentById(id);
  const allowed = ['name', 'code', 'equipmentType', 'status', 'templateName', 'address', 'instanceNumber'];
  const update = {};
  for (const key of allowed) {
    if (data[key] !== undefined) {
      if (key === 'templateName' || key === 'address' || key === 'instanceNumber') {
        const v = data[key];
        update[key] =
          v === null || v === '' ? null : typeof v === 'string' ? v.trim() : v;
      } else {
        update[key] =
          typeof data[key] === 'string' ? data[key].trim() : data[key];
      }
    }
  }
  if (Object.keys(update).length === 0) {
    throw new HttpError(400, 'No fields to update');
  }
  return prisma.equipment.update({
    where: { id },
    data: update,
  });
}

async function deleteEquipment(id) {
  const existing = await getEquipmentById(id);
  await prisma.equipment.delete({
    where: { id },
  });
  return existing;
}

module.exports = {
  listEquipmentByFloor,
  listEquipmentBySite,
  createEquipment,
  getEquipmentById,
  updateEquipment,
  deleteEquipment,
};

File: server/src/main/java/com/legioncontrols/server/sites/EquipmentService.java
package com.legioncontrols.server.sites;

import com.legioncontrols.server.common.ApiException;
import com.legioncontrols.server.jooq.generated.tables.records.EquipmentRecord;
import com.legioncontrols.server.jooq.generated.tables.records.FloorRecord;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import org.jooq.DSLContext;
import org.jooq.Field;
import org.springframework.stereotype.Service;

import static com.legioncontrols.server.jooq.generated.tables.Building.BUILDING;
import static com.legioncontrols.server.jooq.generated.tables.Equipment.EQUIPMENT;
import static com.legioncontrols.server.jooq.generated.tables.Floor.FLOOR;
import static com.legioncontrols.server.jooq.generated.tables.Point.POINT;
import static com.legioncontrols.server.jooq.generated.tables.Site.SITE;
import static org.jooq.impl.DSL.count;

/** Java equivalent of backend/src/modules/equipment/equipment.service.js. */
@Service
public class EquipmentService {

    private final DSLContext dsl;

    public EquipmentService(DSLContext dsl) {
        this.dsl = dsl;
    }

    private FloorRecord assertFloorExists(String floorId) {
        FloorRecord floor = dsl.selectFrom(FLOOR).where(FLOOR.ID.eq(floorId)).fetchOne();
        if (floor == null) {
            throw ApiException.notFound("Floor not found");
        }
        return floor;
    }

    public List<EquipmentDto> listEquipmentByFloor(String floorId) {
        assertFloorExists(floorId);
        return dsl.select(EQUIPMENT, count(POINT.ID))
            .from(EQUIPMENT)
            .leftJoin(POINT).on(POINT.EQUIPMENTID.eq(EQUIPMENT.ID))
            .where(EQUIPMENT.FLOORID.eq(floorId))
            .groupBy(EQUIPMENT.fields())
            .orderBy(EQUIPMENT.NAME.asc())
            .fetch()
            .map(r -> EquipmentDto.from(r.value1(), null, null, r.value2().longValue()));
    }

    public List<EquipmentDto> listEquipmentBySite(String siteId) {
        boolean siteExists = dsl.fetchExists(dsl.selectFrom(SITE).where(SITE.ID.eq(siteId)));
        if (!siteExists) {
            throw ApiException.notFound("Site not found");
        }
        Field<Integer> pointCountField = count(POINT.ID).as("pointCount");
        return dsl.select(EQUIPMENT.fields())
            .select(BUILDING.ID, BUILDING.NAME, FLOOR.ID, FLOOR.NAME, pointCountField)
            .from(EQUIPMENT)
            .leftJoin(BUILDING).on(BUILDING.ID.eq(EQUIPMENT.BUILDINGID))
            .leftJoin(FLOOR).on(FLOOR.ID.eq(EQUIPMENT.FLOORID))
            .leftJoin(POINT).on(POINT.EQUIPMENTID.eq(EQUIPMENT.ID))
            .where(EQUIPMENT.SITEID.eq(siteId))
            .groupBy(groupFields(EQUIPMENT.fields(), BUILDING.ID, BUILDING.NAME, FLOOR.ID, FLOOR.NAME))
            .orderBy(EQUIPMENT.BUILDINGID.asc(), EQUIPMENT.FLOORID.asc(), EQUIPMENT.NAME.asc())
            .fetch()
            .map(r -> {
                EquipmentRecord eq = r.into(EQUIPMENT);
                EquipmentDto.RefDto building = r.get(BUILDING.ID) != null ? new EquipmentDto.RefDto(r.get(BUILDING.ID), r.get(BUILDING.NAME)) : null;
                EquipmentDto.RefDto floor = r.get(FLOOR.ID) != null ? new EquipmentDto.RefDto(r.get(FLOOR.ID), r.get(FLOOR.NAME)) : null;
                Integer pointCount = r.get(pointCountField);
                return EquipmentDto.from(eq, building, floor, pointCount != null ? pointCount.longValue() : 0L);
            });
    }

    public EquipmentRecord getEquipmentRecordById(String id) {
        EquipmentRecord equipment = dsl.selectFrom(EQUIPMENT).where(EQUIPMENT.ID.eq(id)).fetchOne();
        if (equipment == null) {
            throw ApiException.notFound("Equipment not found");
        }
        return equipment;
    }

    public record CreateEquipmentRequest(String name, String code, String equipmentType, String status,
                                          String templateName, String address, String instanceNumber) {
    }

    public EquipmentRecord createEquipment(String floorId, CreateEquipmentRequest req) {
        FloorRecord floor = assertFloorExists(floorId);
        String buildingId = floor.getBuildingid();
        String siteId = dsl.select(BUILDING.SITEID).from(BUILDING).where(BUILDING.ID.eq(buildingId)).fetchOne(BUILDING.SITEID);

        if (isBlank(req.name()) || isBlank(req.code()) || isBlank(req.equipmentType())) {
            throw ApiException.badRequest("name, code, and equipmentType are required");
        }

        EquipmentRecord record = dsl.newRecord(EQUIPMENT);
        record.setId(UUID.randomUUID().toString());
        record.setSiteid(siteId);
        record.setBuildingid(buildingId);
        record.setFloorid(floorId);
        record.setName(req.name().trim());
        record.setCode(req.code().trim());
        record.setEquipmenttype(req.equipmentType().trim());
        record.setTemplatename(blankToNull(req.templateName()));
        if (req.status() != null) {
            record.setStatus(com.legioncontrols.server.jooq.generated.enums.Entitystatus.valueOf(req.status().toUpperCase()));
        }
        record.setAddress(blankToNull(req.address()));
        record.setInstancenumber(blankToNull(req.instanceNumber()));
        record.setCreatedat(LocalDateTime.now());
        record.setUpdatedat(LocalDateTime.now());
        record.insert();
        record.refresh();
        return record;
    }

    public record UpdateEquipmentRequest(String name, String code, String equipmentType, String status,
                                          String templateName, String address, String instanceNumber) {
    }

    public EquipmentRecord updateEquipment(String id, UpdateEquipmentRequest req) {
        EquipmentRecord record = getEquipmentRecordById(id);
        boolean changed = false;
        if (req.name() != null) { record.setName(req.name().trim()); changed = true; }
        if (req.code() != null) { record.setCode(req.code().trim()); changed = true; }
        if (req.equipmentType() != null) { record.setEquipmenttype(req.equipmentType().trim()); changed = true; }
        if (req.status() != null) { record.setStatus(com.legioncontrols.server.jooq.generated.enums.Entitystatus.valueOf(req.status().toUpperCase())); changed = true; }
        if (req.templateName() != null) { record.setTemplatename(blankToNull(req.templateName())); changed = true; }
        if (req.address() != null) { record.setAddress(blankToNull(req.address())); changed = true; }
        if (req.instanceNumber() != null) { record.setInstancenumber(blankToNull(req.instanceNumber())); changed = true; }
        if (!changed) {
            throw ApiException.badRequest("No fields to update");
        }
        record.setUpdatedat(LocalDateTime.now());
        record.update();
        return record;
    }

    public EquipmentRecord deleteEquipment(String id) {
        EquipmentRecord record = getEquipmentRecordById(id);
        dsl.deleteFrom(EQUIPMENT).where(EQUIPMENT.ID.eq(id)).execute();
        return record;
    }

    private static org.jooq.GroupField[] groupFields(org.jooq.Field<?>[] tableFields, org.jooq.GroupField... extra) {
        org.jooq.GroupField[] combined = new org.jooq.GroupField[tableFields.length + extra.length];
        System.arraycopy(tableFields, 0, combined, 0, tableFields.length);
        System.arraycopy(extra, 0, combined, tableFields.length, extra.length);
        return combined;
    }

    private static boolean isBlank(String s) {
        return s == null || s.isBlank();
    }

    private static String blankToNull(String s) {
        if (s == null) return null;
        String trimmed = s.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }
}
